### PR TITLE
add new features in `scale_to()` so user can run without an x-value

### LINKED ISSUE #287: DiffractionObject.scale_to() default is to scale to maxima
### Problem
Sorry guys, this is a scope creep, but I had an a idea to make the `scale_to()` method more useful so that it can be run with no xvalue passed and still work by default. 

### Proposed solution
 It would be something like this:
```
scaled_do = my_do.scaled_to(target_do)
```
and the default scaling would simply be `max(target_do.on_q()[0])/max(my_do.on_q()[0])`

if user passes in an xvalue, and xtype we go to the current workflow.  This allows the user to override the default behavior when it doesn't work for some reason, for example they want to scale to a diffraction feature but there is small angle scattering in the signal, or the maximum peak in one pattern is a different peak or something.

### PR DESCRIPTION
**Pull request by yucongalicechen**
closes #287
@sbillinge reafy for review

### CODE REVIEW

File: src/diffpy/utils/diffraction_objects.py [hunk 1]
```diff
@@ -402,14 +402,16 @@ def scale_to(self, target_diff_object, q=None, tth=None, d=None, offset=0):
 
         The y-value in the target at the closest specified x-value will be used as the factor to scale to.
         The entire array is scaled by this factor so that one object places on top of the other at that point.
-        If multiple values of `q`, `tth`, or `d` are provided, or none are provided, an error will be raised.
+        If none of `q`, `tth`, or `d` are provided,
+        the scaling will be based on the maximal x-array value from both objects.
```
Comment (yucongalicechen) [inline src/diffpy/utils/diffraction_objects.py:406]: add new feature

File: src/diffpy/utils/diffraction_objects.py [hunk 2]
```diff
@@ -402,14 +402,16 @@ def scale_to(self, target_diff_object, q=None, tth=None, d=None, offset=0):
...
+        If multiple values of `q`, `tth`, or `d` are provided, an error will be raised.
 
         Parameters
         ----------
         target_diff_object: DiffractionObject
             the diffraction object you want to scale the current one onto
 
-        q, tth, d : float, optional, must specify exactly one of them
+        q, tth, d : float, optional, default is q with the maximal x-array value of the current object
```
Comment (yucongalicechen) [inline src/diffpy/utils/diffraction_objects.py:414]: edit description

File: src/diffpy/utils/diffraction_objects.py [hunk 3]
```diff
@@ -422,11 +424,16 @@ def scale_to(self, target_diff_object, q=None, tth=None, d=None, offset=0):
         """
         scaled = self.copy()
         count = sum([q is not None, tth is not None, d is not None])
-        if count != 1:
+        if count > 1:
             raise ValueError(
-                "You must specify exactly one of 'q', 'tth', or 'd'. Please rerun specifying only one."
+                "You must specify none or exactly one of 'q', 'tth', or 'd'. "
```
Comment (yucongalicechen) [inline src/diffpy/utils/diffraction_objects.py:429]: edit error msg

File: src/diffpy/utils/diffraction_objects.py [hunk 4]
```diff
@@ -422,11 +424,16 @@ def scale_to(self, target_diff_object, q=None, tth=None, d=None, offset=0):
...
+                "Please provide either none or one value."
             )
 
+        if count == 0:
+            scaled._all_arrays[:, 0] *= max(target_diff_object.on_q()[1]) / max(self.on_q()[1])
+            return scaled
```
Comment (yucongalicechen) [inline src/diffpy/utils/diffraction_objects.py:435]: when no xvalue was specified

File: tests/test_diffraction_objects.py [hunk 5]
```diff
@@ -284,23 +284,7 @@ def test_init_invalid_xtype():
...
-    # Check the intensity data is the same as expected
-    assert np.allclose(scaled_do.on_xtype(expected["xtype"])[1], expected["yarray"])
-
-
-@pytest.mark.parametrize(
-    "org_do_args, target_do_args, scale_inputs",
-    [
-        # Test expected errors produced from scale_to() with invalid inputs
-        (  # C1: none of q, tth, d, provided, expect ValueError
+        (  # C5: none of q, tth, d, provided, expect to scale on the maximal x-arrays
```
Comment (yucongalicechen) [inline tests/test_diffraction_objects.py:287]: new test case

File: doc/source/examples/diffraction_objects_example.rst [hunk 6]
```diff
@@ -117,6 +117,13 @@ For convenience, you can also apply an offset to the scaled new diffraction obje
 
     scaled_and_offset_measured = measured.scale_to(calculated, q=5.5, offset=0.5)
 
+You can call `scale_to()` without specifying a value for `q`, `tth`, or `d`.
```
Comment (sbillinge) [inline doc/source/examples/diffraction_objects_example.rst:120]: This will be the default behavior so should go first in the docs.    Then we can say something like "if this doesnlt give desirable results you can user....." and describe the other workflow.

To give even more flexibility to the user we may want to allow them to specify a separate xtype and xvalue for the target and the scaled DO.  The hierarchy woiuld be:
1. no xtype/xvalue
2. if that doesn't give desirable results, specify xtype and xvalue for target.  scale to the same (nearest) point in the scaled pattern
3. if that doesn't give desirable results, specify xtype, and a different xvalue for target. and scaled DO, respectively.  Scale the yvale of xvalue_scaled to yvalue of xvalue_target.

This last functionality is a new one and we should put it on a new issue (and a new PR if we want to implement it)

File: news/scaleto-max.rst [hunk 7]
```diff
@@ -0,0 +1,23 @@
+**Added:**
+
+* new feature in `scale_to()` to run without specifying an x-value
```
Comment (sbillinge) [inline news/scaleto-max.rst:3]: describe what it does, not what it doesn't do...i.e., scales to max value in each chase.  You can also add that it allows this to be used without xvalue

File: src/diffpy/utils/diffraction_objects.py (hunk 1 above)
Comment (sbillinge) [reply src/diffpy/utils/diffraction_objects.py:406]: as above.  Put this behavior first.

File: tests/test_diffraction_objects.py (hunk 5 above)
Comment (sbillinge) [reply tests/test_diffraction_objects.py:287]: put this one first?  I think it is worth the extra effort as this is our new default behavior.

File: tests/test_diffraction_objects.py [hunk 8]
```diff
@@ -319,8 +303,25 @@ def test_scale_to(org_do_args, target_do_args, scale_inputs, expected):
                 "d": None,
                 "offset": 0,
```
Comment (sbillinge) [inline tests/test_diffraction_objects.py:304]: this is a different issue, unrelated to what you are doing here, but I would rather have default offset to be None (set to zero if None).  I think this is more logical for my brain.  If I don't set an offset, the offset is None.

File: tests/test_diffraction_objects.py [hunk 9]
```diff
@@ -319,8 +303,25 @@ def test_scale_to(org_do_args, target_do_args, scale_inputs, expected):
...
+            {"xtype": "q", "yarray": np.array([10, 20, 30])},
         ),
-        (  # C2: tth and d both provided, expect ValueErrort
+    ],
+)
+def test_scale_to(org_do_args, target_do_args, scale_inputs, expected):
+    original_do = DiffractionObject(**org_do_args)
+    target_do = DiffractionObject(**target_do_args)
+    scaled_do = original_do.scale_to(
+        target_do, q=scale_inputs["q"], tth=scale_inputs["tth"], d=scale_inputs["d"], offset=scale_inputs["offset"]
```
Comment (sbillinge) [inline tests/test_diffraction_objects.py:314]: I think we need to fgure out a way only pass in what is needed, rather than passing in everything with some things set to default values.  It may need more test functions written, but see if you can figure out how to do it with unpacking kwargs that are handed in.

**Review by sbillinge — COMMENTED** [review]
Thanks @yucongalicechen !. please see inline comments.

**Comment by yucongalicechen** [comment]
@sbillinge ready for review

File: doc/source/examples/diffraction_objects_example.rst [hunk 10]
```diff
@@ -106,6 +106,16 @@ we would replace the code above with
 The ``scale_to()`` method returns a new ``DiffractionObject`` which we can assign to a new
 variable and make use of,
 
+The default behavior is to align the objects based on the maximal q-value of each diffraction object,
+so they will align at the intensity at these indices.
+
+.. code-block:: python
+
+    scaled_measured = measured.scale_to(calculated)
+
```
Comment (yucongalicechen) [inline doc/source/examples/diffraction_objects_example.rst:115]: put default behavior first

File: news/scaleto-max.rst [hunk 11]
```diff
@@ -0,0 +1,23 @@
+**Added:**
+
+* new feature in `scale_to()`: default scaling is based on the max q-value in each diffraction object.
```
Comment (yucongalicechen) [inline news/scaleto-max.rst:3]: edited

File: src/diffpy/utils/diffraction_objects.py [hunk 12]
```diff
@@ -396,37 +396,46 @@ def on_tth(self):
     def on_d(self):
         return [self.all_arrays[:, 3], self.all_arrays[:, 0]]
 
-    def scale_to(self, target_diff_object, q=None, tth=None, d=None, offset=0):
+    def scale_to(self, target_diff_object, q=None, tth=None, d=None, offset=None):
```
Comment (yucongalicechen) [inline src/diffpy/utils/diffraction_objects.py:399]: change default offset to None

File: src/diffpy/utils/diffraction_objects.py [hunk 13]
```diff
@@ -396,37 +396,46 @@ def on_tth(self):
...
         """Returns a new diffraction object which is the current object but
         rescaled in y to the target.
 
-        The y-value in the target at the closest specified x-value will be used as the factor to scale to.
+        By default, if `q`, `tth`, or `d` are not provided, scaling is based on the max q-value from each object.
+        Otherwise, y-value in the target at the closest specified x-value will be used as the factor to scale to.
```
Comment (yucongalicechen) [inline src/diffpy/utils/diffraction_objects.py:404]: put default behavior first in docstring

File: src/diffpy/utils/diffraction_objects.py (hunk 4 above)
Comment (yucongalicechen) [reply src/diffpy/utils/diffraction_objects.py:437]: If user specified an offset do we want to add offset to the scaled array?

File: tests/test_diffraction_objects.py [hunk 14]
```diff
@@ -276,22 +277,15 @@ def test_init_invalid_xtype():
...
         ),
     ],
 )
 def test_scale_to(org_do_args, target_do_args, scale_inputs, expected):
     original_do = DiffractionObject(**org_do_args)
     target_do = DiffractionObject(**target_do_args)
-    scaled_do = original_do.scale_to(
-        target_do, q=scale_inputs["q"], tth=scale_inputs["tth"], d=scale_inputs["d"], offset=scale_inputs["offset"]
-    )
+    scaled_do = original_do.scale_to(target_do, **scale_inputs)
```
Comment (yucongalicechen) [inline tests/test_diffraction_objects.py:288]: edited - this allows us to only pass in what we need

File: src/diffpy/utils/diffraction_objects.py [hunk 15]
```diff
@@ -396,32 +396,34 @@ def on_tth(self):
...
-        the scaling will be based on the maximal x-array value from both objects.
         If multiple values of `q`, `tth`, or `d` are provided, an error will be raised.
 
         Parameters
         ----------
         target_diff_object: DiffractionObject
             the diffraction object you want to scale the current one onto
 
-        q, tth, d : float, optional, default is q with the maximal x-array value of the current object
+        q, tth, d : float, optional, default is the max q-value from each object
```
Comment (sbillinge) [inline src/diffpy/utils/diffraction_objects.py:413]: "default is None"

File: src/diffpy/utils/diffraction_objects.py [hunk 16]
```diff
@@ -396,32 +396,34 @@ def on_tth(self):
...
             The value of the x-array where you want the curves to line up vertically.
             Specify a value on one of the allowed grids, q, tth, or d), e.g., q=10.
 
-        offset : float, optional, default is 0
+        offset : float, optional, default is None
             an offset to add to the scaled y-values
```
Comment (sbillinge) [inline src/diffpy/utils/diffraction_objects.py:418]: "The offset to be added to the scaled y-values"

File: src/diffpy/utils/diffraction_objects.py [hunk 17]
```diff
@@ -396,32 +396,34 @@ def on_tth(self):
...
 
         Returns
         -------
         the rescaled DiffractionObject as a new object
         """
+        if offset is None:
+            offset = 0
+
```
Comment (sbillinge) [inline src/diffpy/utils/diffraction_objects.py:426]: remove empty line

**Review by sbillinge — COMMENTED** [review]
small tweaks

File: src/diffpy/utils/diffraction_objects.py (hunk 4 above)
Comment (sbillinge) [reply src/diffpy/utils/diffraction_objects.py:437]: yes, that is correct.  Add the offset to the y-array after scaling

File: tests/test_diffraction_objects.py [hunk 18]
```diff
@@ -255,15 +261,10 @@ def test_init_invalid_xtype():
...
-                "offset": 0,
-            },
+            {"tth": 60},
             {"xtype": "tth", "yarray": np.array([1, 2, 3, 4, 5, 6, 10])},
         ),
-        (  # C4: Same x-array and y-array with 2.1 offset, expect y-array to shift up by 2.1
+        (  # C5: Same x-array and y-array with 2.1 offset, expect y-array to shift up by 2.1
             {
                 "xarray": np.array([10, 15, 25, 30, 60, 140]),
                 "yarray": np.array([2, 3, 4, 5, 6, 7]),
```
Comment (sbillinge) [inline tests/test_diffraction_objects.py:270]: This test is compromised because the target and scale y-arrays are the same, so it is and edge case.  In general, we want to scale, then offset.

**Review by sbillinge — COMMENTED** [review]
please see inline.  I answered a question, and raised a new one.

File: src/diffpy/utils/diffraction_objects.py [hunk 19]
```diff
@@ -396,37 +396,48 @@ def on_tth(self):
     def on_d(self):
         return [self.all_arrays[:, 3], self.all_arrays[:, 0]]
 
-    def scale_to(self, target_diff_object, q=None, tth=None, d=None, offset=0):
+    def scale_to(self, target_diff_object, q=None, tth=None, d=None, offset=None):
         """Returns a new diffraction object which is the current object but
         rescaled in y to the target.
 
-        The y-value in the target at the closest specified x-value will be used as the factor to scale to.
+        By default, if `q`, `tth`, or `d` are not provided, scaling is based on the max intensity from each object.
```
Comment (yucongalicechen) [inline src/diffpy/utils/diffraction_objects.py:403]: change max q-value to max intensity here

File: src/diffpy/utils/diffraction_objects.py (hunk 4 above)
Comment (yucongalicechen) [reply src/diffpy/utils/diffraction_objects.py:439]: added the offset here

File: tests/test_diffraction_objects.py [hunk 20]
```diff
@@ -255,43 +261,48 @@ def test_init_invalid_xtype():
...
-                "tth": 60,
-                "d": None,
-                "offset": 0,
-            },
+            {"tth": 60},
             {"xtype": "tth", "yarray": np.array([1, 2, 3, 4, 5, 6, 10])},
         ),
-        (  # C4: Same x-array and y-array with 2.1 offset, expect y-array to shift up by 2.1
+        (  # C5.1: Reuse test case from C1, none of q, tth, d, provided, but include an offset,
+            # expect scaled y-array in C1 to shift up by 2
```
Comment (yucongalicechen) [inline tests/test_diffraction_objects.py:268]: add tests for offset, reuse patterns from C1 and C4 so that they're more general

**Comment by yucongalicechen** [comment]
@sbillinge ready for review!
